pkg/controller.go

```diff
@@ -127,46 +144,101 @@ func (c *controller) Run(ctx context.Context) error {
-		}
-		if c.opts.DryRun {
-			evictionPolicy.DeleteOptions = &metav1.DeleteOptions{
-				DryRun: []string{"All"},
-			}
-		}
-		klog.V(1).Infof("Evicting Pod '%s/%s'", pod.Namespace, pod.Name)
-		c.recorder.Event(pod.DeepCopyObject(), "Normal", "SoftEviction", fmt.Sprintf("Pod '%s/%s' has at least one container close to its memory limit", pod.Namespace, pod.Name))
-
-		err := c.clientset.PolicyV1beta1().Evictions(pod.Namespace).Evict(ctx, &evictionPolicy)
+// evict pods with a pause between each eviction
+func (c *controller) evictWithPauseChanLoop(ctx context.Context) {
+	for pod := range c.pauseChan {
+		err := c.evictPod(ctx, pod)
 		if err != nil {
 			c.recorder.Event(pod.DeepCopyObject(), "Warning", "SoftEviction", fmt.Sprintf("Unable to evict Pod '%s/%s' : %v", pod.Namespace, pod.Name, err))
 			klog.Errorf("error evicting '%s/%s': %v", pod.Namespace, pod.Name, err)
 			continue
 		}
 
 		if c.opts.EvictionPause > 0 {
+			klog.V(2).Infof("Pausing pod eviction for %s", c.opts.EvictionPause)
 			time.Sleep(c.opts.EvictionPause)
 		}
+
+		klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
 	}
-	return nil
 }
 
-func (c *controller) listPodsCloseToMemoryLimit(ctx context.Context) ([]*corev1.Pod, error) {
+// evict pods having a PodDisruptionBudget. If the PDB is exceeded, the eviction
+// is retried after a delay using a different channel.
+func (c *controller) evictWithPDBChanLoop(ctx context.Context) {
+	backoffChan := make(chan *corev1.Pod, c.opts.ChannelBufferSize)
+	defer close(backoffChan)
+	go func() {
+		for pod := range backoffChan {
+			attempt := 1
+		retry:
+			klog.V(2).Infof("Retrying eviction of '%s/%s' (attempt %d)", pod.Namespace, pod.Name, attempt)
+			err := c.evictPod(ctx, pod)
+			if err != nil {
+				if strings.Contains(err.Error(), "disruption budget") && attempt <= 3 {
+					time.Sleep(time.Duration(math.Pow(5, float64(attempt))) * time.Second)
+					attempt++
+					goto retry
+				}
+				c.recorder.Event(pod.DeepCopyObject(), "Warning", "SoftEviction", fmt.Sprintf("Unable to evict Pod '%s/%s' : %v", pod.Namespace, pod.Name, err))
+				klog.Errorf("error evicting '%s/%s': %v", pod.Namespace, pod.Name, err)
+				continue
+			}
+
+			klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
+		}
+	}()
+
+	for pod := range c.pdbChan {
+		err := c.evictPod(ctx, pod)
+		if err != nil {
+			backoffChan <- pod
+		}
+
+		klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
+	}
+}
+
+func (c *controller) schedulePodEviction(ctx context.Context, pod *corev1.Pod) {
+	klog.V(1).Infof("Evicting Pod '%s/%s'", pod.Namespace, pod.Name)
+	c.recorder.Event(pod.DeepCopyObject(), "Normal", "SoftEviction", fmt.Sprintf("Pod '%s/%s' has at least one container close to its memory limit", pod.Namespace, pod.Name))
+
+	if c.hasDisruptionBudget(ctx, pod) {
+		if len(c.pdbChan) == cap(c.pdbChan) {
+			klog.Warningf("PDB channel is full, skipping eviction for Pod '%s/%s'", pod.Namespace, pod.Name)
```

Comment:
We're not monitoring this controller with metrics. What about a safety net, which would consist in having a "skipped eviction counter", reset when we enqueue, incremented when we skip, and crash the controller if we reach let's say 10 skipped evictions ?

Reply:
In general I like the idea. If the queue is full, it will likely be soon again. The forced restart will empty the queue and maybe even trigger a notification based on frequent restarts.

However, when evictions are being skipped it is because the queue is already full. We won't miss an eviction though, because there will be another run trying to schedule pods to be evicted before the queue has been emptied (at least if you run the controller close to the default values).

I think this could be a dedicated change.